From hidave.darkstar@gmail.com  Tue Feb 17 17:41:08 2009
From: Dave Young <hidave.darkstar@gmail.com>
Date: Sat, 14 Feb 2009 21:21:13 +0800
Subject: usb-serial: fix usb_serial_register bug when boot with nousb param
To: gregkh@suse.de
Message-ID: <4996C549.4040209@gmail.com>


From: Dave Young <hidave.darkstar@gmail.com>

With "nousb" cmdline booting, built-in serial drivers (ie. airecable)
will trigger kernel oops.

Indeed, if nousb, usb_serial_init will failed, and the usb serial bus type
will not be registerd, then usb_serial_register call driver_register
which try to register the driver to a not registered bus.

Here add usb_disabled() check in usb_serial_register to fix it.

Signed-off-by: Dave Young <hidave.darkstar@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/serial/usb-serial.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/usb/serial/usb-serial.c
+++ b/drivers/usb/serial/usb-serial.c
@@ -1241,6 +1241,9 @@ int usb_serial_register(struct usb_seria
 	/* must be called with BKL held */
 	int retval;
 
+	if (usb_disabled())
+		return -ENODEV;
+
 	fixup_generic(driver);
 
 	if (!driver->description)
